src/main/java/com/r307/arbitrader/service/event/StreamingTickerEventListener.java

```diff
@@ -0,0 +1,21 @@
+package com.r307.arbitrader.service.event;
+
+import com.r307.arbitrader.service.TradingService;
+import com.r307.arbitrader.service.model.TickerEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.scheduling.annotation.Async;
+
+
+public class StreamingTickerEventListener {
```

Comment:
I'm not super familiar with the application event model in Spring so I'm doing some reading to get up to speed but maybe you already know the answers to these questions.

* Do the asynchronous event handlers fire simultaneously in different threads? We could be opening up the possibility of some complicated concurrency scenarios in this case.
* Are async events guaranteed to be handled in order?
* Can we configure the event executor to limit it to one executor or increase the size of the queue?
* Can we query the executor to see how big the queue is, clear it, or avoid publishing a new event if there are too many stacked up already?

I'm asking because using async events challenges some of the base assumptions I've made in Arbitrader up to this point. The scheduler guarantees that only one update is running at a time. If it takes more than the 3 seconds we gave it the next update waits to run. That allowed me to avoid any synchronization issues, and conditions like trying to open two pairs of trades in different threads at the same time. If we get a lot of price updates at once and they all trigger trade logic in multiple threads we need to be sure we can block all but the first thread from actually trading.

Reply:
Good questions, I am also not an expert on Spring Event so take my words if a grain of salt.

AFAIK:

_Do the asynchronous event handlers fire simultaneously in different threads? We could be opening up the possibility of some complicated concurrency scenarios in this case._
Yes and this is something I did not think about when taking this approach.

_Are async events guaranteed to be handled in order?_
Yes

_Can we configure the event executor to limit it to one executor or increase the size of the queue?_
Yes, we can define how many threads to use to consume events. To do so we need to provide an `Executor` (where the thread pool size is defined). Increasing the queue size is also possible via the `Executor`. By default the size is integer max value.

_Can we query the executor to see how big the queue is, clear it, or avoid publishing a new event if there are too many stacked up already?_
I thought about the corner case when the consumer is too slow consuming the events but later on I lost the focus on this specific detail and totally forgot about it. Good that you ask. Answering your question, yes it is possible. To do so we need to inject `Executor` and then call it's methods for each functionality you described. For example: `executor.getThreadPoolExecutor().getQueue().clear();`

I opted for Spring Events because it is quick and easy to implement but now I am not so sure if it is the best tool for our use case. Do you have some other tool/framework in mind?